// @thirdweb-dev/wallets/dist/connector-c1e1817a.browser.esm.js
import { _ as _defineProperty } from './defineProperty-c8ecdc07.browser.esm.js';
import EventEmitter from 'eventemitter3';

class Connector extends EventEmitter {}
class WagmiAdapter extends Connector {
  constructor(wagmiConnector) {
    super();
    _defineProperty(this, "wagmiConnector", void 0);
    this.wagmiConnector = wagmiConnector;
  }
  async connect(args) {
    this.setupConnectorListeners();
    const result = await this.wagmiConnector.connect(args);
    return result.account;
  }
  disconnect() {
    this.wagmiConnector.removeAllListeners("connect");
    this.wagmiConnector.removeAllListeners("change");
    return this.wagmiConnector.disconnect();
  }
  isConnected() {
    return this.wagmiConnector.isAuthorized();
  }
  getAddress() {
    return this.wagmiConnector.getAccount();
  }
  getSigner() {
    return this.wagmiConnector.getSigner();
  }
  getProvider() {
    return this.wagmiConnector.getProvider();
  }
  async switchChain(chainId) {
    if (!this.wagmiConnector.switchChain) {
      throw new Error("Switch chain not supported");
    }
    await this.wagmiConnector.switchChain(chainId);
  }
  setupConnectorListeners() {
    this.wagmiConnector.addListener("connect", data => {
      this.emit("connect", data);
    });
    this.wagmiConnector.addListener("change", data => {
      this.emit("change", data);
    });
    this.wagmiConnector.addListener("disconnect", () => {
      this.emit("disconnect");
    });
  }
  async setupListeners() {
    this.setupConnectorListeners();
    await this.wagmiConnector.setupListeners();
  }
  updateChains(chains) {
    this.wagmiConnector.updateChains(chains);
  }
}

export { Connector as C, WagmiAdapter as W };
