// @thirdweb-dev/wallets/dist/thirdweb-dev-wallets.browser.esm.js
export { L as LocalStorage, S as SmartWallet, b as WalletConnectHandler, W as WalletConnectV1Handler, a as WalletConnectV2Handler, c as createLocalStorage, g as getAllSigners, d as getAllSmartWallets, e as getSmartWalletAddress, i as isSmartWalletDeployed } from './smart-wallet-551709d2.browser.esm.js';
export { a as AbstractClientWallet, A as AsyncLocalStorage, D as DEFAULT_DAPP_META, c as createAsyncLocalStorage } from './base-8189a847.browser.esm.js';
export { E as EIP155_SIGNING_METHODS } from './wc-c6a6a61c.browser.esm.js';
export { w as walletIds } from './walletIds-99b78e4a.browser.esm.js';
export { W as WagmiConnector } from './WagmiConnector-60b19604.browser.esm.js';
export { A as AddChainError, C as ChainNotConfiguredError, P as ProviderRpcError, S as SwitchChainError, U as UserRejectedRequestError } from './errors-4bdf81d2.browser.esm.js';
export { n as normalizeChainId } from './normalizeChainId-e4cc0175.browser.esm.js';
export { C as Connector, W as WagmiAdapter } from './connector-c1e1817a.browser.esm.js';
export { a as assertWindowEthereum } from './assertWindowEthereum-66ccef46.browser.esm.js';
export { PaperWallet } from '../evm/wallets/paper-wallet/dist/thirdweb-dev-wallets-evm-wallets-paper-wallet.browser.esm.js';
export { CoinbaseWallet } from '../evm/wallets/coinbase-wallet/dist/thirdweb-dev-wallets-evm-wallets-coinbase-wallet.browser.esm.js';
export { LocalWallet, isValidPrivateKey } from '../evm/wallets/local-wallet/dist/thirdweb-dev-wallets-evm-wallets-local-wallet.browser.esm.js';
export { FrameWallet } from '../evm/wallets/frame/dist/thirdweb-dev-wallets-evm-wallets-frame.browser.esm.js';
export { InjectedWallet } from '../evm/wallets/injected/dist/thirdweb-dev-wallets-evm-wallets-injected.browser.esm.js';
export { MetaMaskWallet } from '../evm/wallets/metamask/dist/thirdweb-dev-wallets-evm-wallets-metamask.browser.esm.js';
export { WalletConnect } from '../evm/wallets/wallet-connect/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect.browser.esm.js';
export { WalletConnectV1 } from '../evm/wallets/wallet-connect-v1/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect-v1.browser.esm.js';
export { S as SafeSupportedChainsSet } from './constants-cfec126a.browser.esm.js';
export { SafeWallet } from '../evm/wallets/safe/dist/thirdweb-dev-wallets-evm-wallets-safe.browser.esm.js';
export { MagicLink } from '../evm/wallets/magic/dist/thirdweb-dev-wallets-evm-wallets-magic.browser.esm.js';
export { EthersWallet } from '../evm/wallets/ethers/dist/thirdweb-dev-wallets-evm-wallets-ethers.browser.esm.js';
export { PrivateKeyWallet } from '../evm/wallets/private-key/dist/thirdweb-dev-wallets-evm-wallets-private-key.browser.esm.js';
export { ZerionWallet } from '../evm/wallets/zerion/dist/thirdweb-dev-wallets-evm-wallets-zerion.browser.esm.js';
import './classPrivateMethodGet-ea199cc3.browser.esm.js';
import './classPrivateFieldSet-a5db7c83.browser.esm.js';
import './defineProperty-c8ecdc07.browser.esm.js';
import '@walletconnect/core';
import '@walletconnect/web3wallet';
import 'ethers';
import 'eventemitter3';
import '@walletconnect/jsonrpc-utils';
import '@walletconnect/client';
import '@thirdweb-dev/sdk';
import '../evm/wallets/abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.browser.esm.js';
import '@thirdweb-dev/chains';
import 'buffer';
import './keys-bd983fca.browser.esm.js';

class CredentialsStorage {
  async getItem() {
    const credential = await navigator.credentials.get({
      password: true,
      unmediated: true
    });
    if (credential && "password" in credential) {
      return credential.password;
    }
    return null;
  }
  async setItem(key, value) {
    const credential = await navigator.credentials.create({
      password: {
        id: key,
        name: key,
        password: value
      }
    });
    if (!credential) {
      throw new Error("Credential not created");
    }
    await navigator.credentials.store(credential);
  }
  async removeItem() {
    // Question: is there any way to remove a credential?
    const credential = await navigator.credentials.get({
      password: true,
      unmediated: true
    });
    if (credential) {
      await navigator.credentials.preventSilentAccess();
    }
  }
}

export { CredentialsStorage };
